### PR TITLE
update keras backend to accommodate Tensor API changes (shape/rank)

### PR DESCRIPTION
**Pull request by mars1026**
The Tensor class used to have a shape attribute; this updates the keras backend to use the rank or dtype properties instead, or the compute_shape() method.

### CODE REVIEW

File: plaidml/bridge/keras/__init__.py [hunk 1]
```diff
@@ -1535,11 +1537,13 @@ def softsign(x):
 
 @_log_call
 def sparse_categorical_crossentropy(target, output, from_logits=False):
-    dims = edsl.TensorDims(output.tensor.shape.ndims)
+    shape = output.tensor.compute_shape()
+    dims = edsl.TensorDims(shape.rank)
```
Comment (tzerrell) [inline plaidml/bridge/keras/__init__.py]: I think this should just be `dims = edsl.TensorDims(output.tensor.rank)`

File: plaidml/bridge/keras/__init__.py [hunk 2]
```diff
@@ -679,7 +679,7 @@ def conv3d_transpose(x,
 @_log_call
 def count_params(x):
     result = 1
-    for dim in x.tensor.shape.int_dims:
+    for dim in x.tensor.compute_shape().into_TensorShape().sizes:
```
Comment (flaub) [inline plaidml/bridge/keras/__init__.py]: I don't think we need to convert to `TensorShape` here, but `compute_shape()` is probably necessary.

File: plaidml/bridge/keras/__init__.py (hunk 1 above)
Comment (flaub) [reply plaidml/bridge/keras/__init__.py]: Agreed, avoiding `compute_shape()` is key because its a fairly expensive operation. (which was the motivation for the API change to begin with actually).

File: plaidml/bridge/keras/__init__.py (hunk 2 above)
Comment (flaub) [reply plaidml/bridge/keras/__init__.py]: Is it possible to use `x.bind_dims()` here instead and compute the product symbolically I wonder?

File: plaidml/bridge/keras/__init__.py (hunk 1 above)
Comment (mars1026) [reply plaidml/bridge/keras/__init__.py]: We can't avoid compute_shape() in this case, because one of the sizes gets passed along in the call one_hot() a couple lines further down; I thought it made more sense to grab the shape object once and reuse it.

Comment (flaub) [reply plaidml/bridge/keras/__init__.py]: Is it possible for `one_hot` to take the dim size symbolically?

File: plaidml/bridge/keras/__init__.py (hunk 2 above)
Comment (tzerrell) [reply plaidml/bridge/keras/__init__.py]: Keras expects `count_params` to return something that's more-or-less a python integer (specifically, going by the uses in the Keras source code, if `n` is a python integer, then `n == X.count_params()` needs to behave as expected and `int(X.count_params())` needs to behave as expected). You could do something like
```
shape = edsl.TensorDims(I.rank)
I.bind_dims(*shape)
ret = 1
for dim in shape:
    ret *= dim
return ret
```
but will that cast and compare to python ints correctly?

Comment (flaub) [reply plaidml/bridge/keras/__init__.py]: OK, no, we have to call` compute_shape()` then.

File: plaidml/bridge/keras/__init__.py (hunk 1 above)
Comment (tzerrell) [reply plaidml/bridge/keras/__init__.py]: `one_hot` needs to construct a tensor equal to `range(this_size)` where `this_size` is equal to the variable we're talking about. I don't see an immediate way to do this symbolically.

This is in `sparse_categorical_crossentropy` and we haven't yet prioritized handling the sparsity for this function in a performant way at runtime, so I'm not sure bad compile time perf matters that much in this case.
